ansible/roles/test/files/ptftests/py3/ip_in_ip_tunnel_test.py

```diff
@@ -216,19 +241,47 @@ def send_and_verify_packets(self):
+                        tunnel_pkt = self.generate_expected_packet(inner_pkt)
+                        l3packet = inner_pkt.getlayer(IP) or inner_pkt.getlayer(IPv6)
+                        self.logger.info("Sending packet dst_mac = {} src_mac = {} dst_ip = {} src_ip = {} from port {}" \
+                            .format(inner_pkt[Ether].dst, inner_pkt[Ether].src, l3packet.dst, l3packet.src, src_port))
+                        self.dataplane.flush()
+                        send_packet(self, src_port, inner_pkt)
+                        # Verify packet is received from IPinIP tunnel
+                        idx, count = verify_packet_any_port(test=self,
+                                                            pkt=tunnel_pkt,
+                                                            ports=dst_ports,
+                                                            device_number=0,
+                                                            timeout=TIMEOUT)
+                        pkt_distribution[self.indice_to_portchannel[dst_ports[idx]]] = pkt_distribution.get(self.indice_to_portchannel[dst_ports[idx]], 0) + 1
+                    is_balance = self.check_balance(pkt_distribution, hash_key, k+1)
+                    if is_balance:
+                        self.logger.info("After verification for {} times, the traffic is balanced.".format(k+1))
+                        return
+                assert(is_balance)
+            # For other completeness level, just do basic check
+            # if receive any expected packet on every portchannel then pass
+            else:
+                self.logger.info("Verifying traffic on {} completeness level, send {} packets.".format(self.completeness_level, BASIC_PACKET_NUM))
+                for i in range(0, BASIC_PACKET_NUM):
+                    inner_pkt = self.generate_packet_to_server(hash_key)
+                    tunnel_pkt = self.generate_expected_packet(inner_pkt)
+                    l3packet = inner_pkt.getlayer(IP) or inner_pkt.getlayer(IPv6)
+                    self.logger.info("Sending packet dst_mac = {} src_mac = {} dst_ip = {} src_ip = {} from port {}" \
+                        .format(inner_pkt[Ether].dst, inner_pkt[Ether].src, l3packet.dst, l3packet.src, src_port))
+                    self.dataplane.flush()
+                    send_packet(self, src_port, inner_pkt)
+                    # Verify packet is received from IPinIP tunnel
+                    idx, count = verify_packet_any_port(test=self,
+                                                        pkt=tunnel_pkt,
+                                                        ports=dst_ports,
+                                                        device_number=0,
+                                                        timeout=TIMEOUT)
+                    pkt_distribution[self.indice_to_portchannel[dst_ports[idx]]] = pkt_distribution.get(self.indice_to_portchannel[dst_ports[idx]], 0) + 1
```

Comment:
@ZhaohuiS 
there are two things. 

1. maybe 20 pkts is not engough, considering some case, there are up 16 uplinks for some T1, could we use 200 pkts?

2. I think the check_received check condition may need some enhancement.
if all pkts received on only one port. The pkt_distribution could be like

pkt_distribution = {"pc1" : 20}

but it can pass the check below, only one item, and the count is 20, so not_received = True.
    def check_received(self, pkt_distribution, hash_key):
        not_received = True
        for portchannel, count in pkt_distribution.items():
            self.logger.info("%-10s \t %10s \t %10s \t" % (portchannel, ">0", str(count)))
            if count <= 0:
                not_received = False
        if not not_received:
            self.logger.info("Check balance failed for {}".format(hash_key))
        assert(not_received)

Reply:
@StormLiangMS Thanks for the comments.
1. This script only runs on T0 topo, but you are right, 20 may not be enough. I increase it to 100.
2. Good catch! I initiate pkt_distribution before sending packets. 
Please see my latest commit.

Here is the log:
```
01:32:46.459  IPinIPTunnel: INFO    : Init pkt_distribution:{'PortChannel104': 0, 'PortChannel101': 0, 'PortChannel103': 0, 'PortChannel102': 0}
01:32:59.776  IPinIPTunnel: INFO    : PortChannel104             >0              18     
01:32:59.776  IPinIPTunnel: INFO    : PortChannel101             >0              23     
01:32:59.776  IPinIPTunnel: INFO    : PortChannel103             >0              24     
01:32:59.776  IPinIPTunnel: INFO    : PortChannel102             >0              35     
01:32:59.776  root      : INFO    : ** END TEST CASE ip_in_ip_tunnel_test.IpinIPTunnelTest
```